list-objects-filter.c

```diff
@@ -620,6 +621,98 @@ static void filter_object_type__init(
 	filter->free_fn = free;
 }
 
+/*
+ * A filter which passes the objects to a compile-time plugin.
+ * The plugin needs to implement the filter_profile_plugin interface
+ * defined in list-objects-filter.h.
+ */
+
+#ifndef FILTER_PROFILE_PLUGINS
+#define FILTER_PROFILE_PLUGINS
+#endif
+
+static const struct filter_profile_plugin
+	*filter_profile_plugins[] = {
+		FILTER_PROFILE_PLUGINS
```

Comment:
Yeah this is my plan - the Makefile should define -DFILTER_PROFILE_PLUGINS to be something like the following:
`&filter_profile_plugin_rand, &filter_profile_plugin_spatialfilter, &filter_profile_plugin_foobar`

Since starting work on this I now thing I realise I need two pieces of code that change - like the following:
1. Type-declaration: `const struct filter_profile_plugin FILTER_PROFILE_PLUGINS ;`
where FILTER_PROFILE_PLUGINS=`filter_profile_plugin_rand, filter_profile_plugin_spatialfilter`
2. Array definition: `static const struct filter_profile_plugin *filter_profile_plugins[] = { FILTER_PROFILE_PLUGIN_PTRS };`
where FILTER_PROFILE_PLUGIN_PTRS =`&filter_profile_plugin_rand, &filter_profile_plugin_spatialfilter`

I think that these are probably fine as inlined preprocessor macros - (1) could be included in the user-supplied plugin headers that we refer to eg `rand.h`, `spatialfilter.h` but it's a bit strange since then we actually supply the entire definition of what we require to be in that header to the user, f, for them to basically copy and paste, and it's only one line: `const struct filter_profile_plugin filter_profile_plugin_rand;`. (The user could put other things in that file .h if they want to for their own purposes, but we won't use it). So, seems reasonable that we would just not bother with user-supplied header files at all since we already know what type everything we want from the user is.

(2) on the other hand, should be in a .c file - not a header file. Again it's only a single line - the list of every user plugin. The user shouldn't define it for us - it should be generated based on what plugins are required by the Makfile arguments. Seems to make just as much sense to put a single line of C in an existing C file using a inlined preprocessor macro, rather than generating a separate file.

Will have a look to see if there is anyway I can do (1) and (2) at once with a single macro, but I probably can't, and it probably doesn't really matter